README.md
# jwk-thumbprint
![npm](https://img.shields.io/npm/v/jwk-thumbprint.svg) [![CircleCI](https://circleci.com/gh/nwtgck/jwk-thumbprint-npm.svg?style=shield)](https://circleci.com/gh/nwtgck/jwk-thumbprint-npm)

JWK Thumbprint for JavaScript/TypeScript on both Web Browser and Node.js

## Installation

```bash
npm i -S jwk-thumbprint
```

## Usage

Here is an usage to reproduce the example in [RFC7638 3.1](<https://tools.ietf.org/html/rfc7638#section-3.1>).

```ts
// TypeScript
// (Remove types to use it in JavaScript)

import {jwkThumbprint} from 'jwk-thumbprint';

const myJwk: JsonWebKey & {kty: 'RSA', kid: string} = {
  kty: 'RSA',
  n: '0vx7agoebGcQSuuPiLJXZptN9nndrQmbXEps2aiAFbWhM78LhWx4cbbfAAtVT86zwu1RK7aPFFxuhDR1L6tSoc_BJECPebWKRXjBZCiFV4n3oknjhMstn64tZ_2W-5JsGY4Hc5n9yBXArwl93lqt7_RN5w6Cf0h4QyQ5v-65YGjQR0_FDW2QvzqY368QQMicAtaSqzs8KJZgnYb9c7d0zgdAZHzu6qMQvRL5hajrn1n91CbOpbISD08qNLyrdkt-bFTWhAI4vMQFh6WeZu0fM4lFd2NcRwr3XPksINHaQ-G_xBniIqbw0Ls1jF44-csFCur-kEgU8awapJzKnqDKgw',
  e: 'AQAB',
  alg: 'RS256',
  kid: '2011-04-29',
};

console.log(jwkThumbprint(myJwk, 'SHA-256'));
// => new Uint8Array([55, 54, 203, 177, 120, 124, 184, 48, 156, 119, 238, 140, 55, 5, 197, 225, 111, 251, 158, 133, 151, 21, 144, 31, 30, 76, 89, 177, 17, 130, 245, 123])
```

### Thumbprints in number array, hex and base64url encodings

First, import the following.

```ts
import {jwkThumbprintByEncoding} from 'jwk-thumbprint';
```

You can get different encodings by `'number'`, `'hex'`, `'base64url'` or `'uint8array'` in **type-safe way**. The return types are properly typed by the encodings.

```ts
const thumbprint: number[] = jwkThumbprintByEncoding(jwk, "SHA-256", 'numbers');
console.log(thumbprint)
// => 55, 54, 203, 177, 120, 124, 184, 48, 156, 119, 238, 140, 55, 5, 197, 225, 111, 251, 158, 133, 151, 21, 144, 31, 30, 76, 89, 177, 17, 130, 245, 123]
```

```ts
const thumbprint: string = jwkThumbprintByEncoding(jwk, "SHA-256", 'hex');
console.log(thumbprint)
// => 3736cbb1787cb8309c77ee8c3705c5e16ffb9e859715901f1e4c59b11182f57b
```


```ts
const thumbprint: string = jwkThumbprintByEncoding(jwk, "SHA-256", 'base64url');
console.log(thumbprint)
// => NzbLsXh8uDCcd-6MNwXF4W_7noWXFZAfHkxZsRGC9Xs";
```

```ts
const thumbprint: Uint8Array = jwkThumbprintByEncoding(jwk, "SHA-256", 'uint8array');
console.log(thumbprint)
// => new Uint8Array([55, 54, 203, 177, 120, 124, 184, 48, 156, 119, 238, 140, 55, 5, 197, 225, 111, 251, 158, 133, 151, 21, 144, 31, 30, 76, 89, 177, 17, 130, 245, 123])
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)

## [Unreleased]

## 0.1.4 - 2020-08-16
### Changed
* Update dependencies

### Fixed
* Sort keys in JSON [#45](https://github.com/nwtgck/jwk-thumbprint-npm/pull/45) by [@mitar](https://github.com/mitar)

[Unreleased]: https://github.com/nwtgck/jwk-thumbprint-npm/compare/v0.1.4...HEAD


test/index.test.ts
import {jwkThumbprint, jwkThumbprintByEncoding} from '../src';
import * as assert from 'power-assert';

describe('jwk-thumbprint', () => {
  it('should return the same thumbprint as an example in RFC7638', () => {
    // (from: https://tools.ietf.org/html/rfc7638#section-3.1)
    const jwk: JsonWebKey & { kty: "RSA", kid: string } = {
      "kty": "RSA",
      "n": "0vx7agoebGcQSuuPiLJXZptN9nndrQmbXEps2aiAFbWhM78LhWx4cbbfAAtVT86zwu1RK7aPFFxuhDR1L6tSoc_BJECPebWKRXjBZCiFV4n3oknjhMstn64tZ_2W-5JsGY4Hc5n9yBXArwl93lqt7_RN5w6Cf0h4QyQ5v-65YGjQR0_FDW2QvzqY368QQMicAtaSqzs8KJZgnYb9c7d0zgdAZHzu6qMQvRL5hajrn1n91CbOpbISD08qNLyrdkt-bFTWhAI4vMQFh6WeZu0fM4lFd2NcRwr3XPksINHaQ-G_xBniIqbw0Ls1jF44-csFCur-kEgU8awapJzKnqDKgw",
      "e": "AQAB",
      "alg": "RS256",
      "kid": "2011-04-29"
    };
    // NOTE: This expectation is also on https://tools.ietf.org/html/rfc7638#section-3.1
    const expectNumbers: number[] = [
      55, 54, 203, 177, 120, 124, 184, 48, 156, 119, 238, 140, 55, 5, 197,
      225, 111, 251, 158, 133, 151, 21, 144, 31, 30, 76, 89, 177, 17, 130,
      245, 123
    ];
    {
      // NOTE: The return-type must not be undefined
      const actual: number[] = jwkThumbprintByEncoding(jwk, "SHA-256", 'numbers');
      assert.deepStrictEqual(actual, expectNumbers);
    }
    {
      // NOTE: The return-type must not be undefined
      const actual: string = jwkThumbprintByEncoding(jwk, "SHA-256", 'hex');
      const expect = '3736cbb1787cb8309c77ee8c3705c5e16ffb9e859715901f1e4c59b11182f57b';
      assert.deepStrictEqual(actual, expect);
    }
    {
      // NOTE: The return-type must not be undefined
      const actual: Uint8Array = jwkThumbprint(jwk, "SHA-256");
      // NOTE: This expectation is also on https://tools.ietf.org/html/rfc7638#section-3.1
      const expect = new Uint8Array(expectNumbers);
      assert.deepStrictEqual(actual, expect);
    }

    {
      // NOTE: The return-type must not be undefined
      const actual: string = jwkThumbprintByEncoding(jwk, "SHA-256", 'base64url');
      // NOTE: This expectation is also in https://tools.ietf.org/html/rfc7638#section-3.1
      const expect = "NzbLsXh8uDCcd-6MNwXF4W_7noWXFZAfHkxZsRGC9Xs";
      assert.deepStrictEqual(actual, expect);
    }
  });

  it('should return symmetric key thumbprint', () => {
    const jwk: JsonWebKey & { kty: "oct" } = {
      "alg": "A128GCM",
      "ext": true,
      "k": "9wcPr5BVF6hku5Fx3IrejQ",
      "key_ops": ["encrypt", "decrypt"],
      "kty": "oct"
    };

    // NOTE: The return-type must not be undefined
    const actual: string = jwkThumbprintByEncoding(jwk, "SHA-256", 'base64url');
    // NOTE: This expectation is calculated by a Ruby library, 'json/jwt'
    //
    // require 'json/jwt'
    //
    // jwk = JSON::JWK.new ({
    //   "alg": "A128GCM",
    //   "ext": true,
    //   "k": "9wcPr5BVF6hku5Fx3IrejQ",
    //   "key_ops": ["encrypt", "decrypt"],
    //   "kty": "oct"
    // })
    // puts(jwk.thumbprint)
    const expect = "O0ohsAio8Tj1lGR2SoX3Xa90quibp6j3vSe71e0LXRY";
    assert.deepStrictEqual(actual, expect);
  });
});


src/index.ts
import * as hashJs from 'hash.js';
import { Base64 } from "js-base64"

export function canonicalizeJwk(jwk: JsonWebKey & {kty: "RSA"}): JsonWebKey;
export function canonicalizeJwk(jwk: JsonWebKey & {kty: "EC"}): JsonWebKey;
export function canonicalizeJwk(jwk: JsonWebKey & {kty: "oct"}): JsonWebKey;
export function canonicalizeJwk(jwk: JsonWebKey): JsonWebKey | undefined;

/**
 * Keep only required members
 * https://tools.ietf.org/html/rfc7638#section-3.1
 * @param jwk
 */
export function canonicalizeJwk(jwk: JsonWebKey): JsonWebKey | undefined {
  switch (jwk.kty) {
    case "RSA":
      return {
        e: jwk.e,
        kty: jwk.kty,
        n: jwk.n,
      };
    case "EC":
      return {
        crv: jwk.crv,
        kty: jwk.kty,
        x: jwk.x,
        y: jwk.y
      };
    case "oct":
      return {
        k: jwk.k,
        kty: jwk.kty,
      };
    default:
      return undefined;
  }
}

// NOTE: these strings are compatible with https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest
type HashAlg = "SHA-256" | "SHA-512";

type Encodings = 'numbers' | 'hex' | 'uint8array' | 'base64url';

type EcType = {
  numbers: number[],
  hex: string;
  uint8array: Uint8Array,
  base64url: string
}

export function jwkThumbprintByEncoding<Ec extends Encodings>(jwk: JsonWebKey & {kty: "RSA" | "EC" | "oct"}, hashAlg: HashAlg, ec: Ec): EcType[Ec];
export function jwkThumbprintByEncoding<Ec extends Encodings>(jwk: JsonWebKey, hashAlg: HashAlg, ec: Ec): EcType[Ec] | undefined;

/**
 * Calculate JWK Thumbprint by encoding
 *
 * https://tools.ietf.org/html/rfc7638#section-3.1
 * @param jwk
 * @param hashAlg
 * @param ec
 */
export function jwkThumbprintByEncoding<Ec extends Encodings>(jwk: JsonWebKey, hashAlg: HashAlg, ec: Ec): EcType[Ec] | undefined {
  // Canonicalize JWK
  const canonicalJwk = canonicalizeJwk(jwk);
  if (canonicalJwk === undefined) {
    return undefined;
  }

  // JSON string sorted by keys
  // (from: https://stackoverflow.com/a/16168003/2885946)
  const jsonStr = JSON.stringify(canonicalJwk, Object.keys(canonicalJwk).sort());

  const digest: Sha256 | Sha512 = (() => {
    switch (hashAlg) {
      case "SHA-256":
        return hashJs.sha256().update(jsonStr);
      case "SHA-512":
        return hashJs.sha512().update(jsonStr);
      default:
        // Never call if the type is valid
        throw new Error(`Unexpected error: unknown algorithm: ${hashAlg}`);
    }
  })();

  switch (ec) {
    case 'numbers':
      return digest.digest();
    case 'hex':
      return digest.digest('hex');
    case "uint8array":
      return new Uint8Array(digest.digest());
    case 'base64url':
      // (from: https://paulownia.hatenablog.com/entry/2019/02/07/201320)
      const binStr = digest.digest().map(b => String.fromCharCode(b)).join("");
      // (from:  https://github.com/brianloveswords/base64url/blob/20117777e233fc86ac1286ccbc998bd6c923f149/src/base64url.ts#L25-L27)
      return Base64.btoa(binStr)
        .replace(/=/g, "")
        .replace(/\+/g, "-")
        .replace(/\//g, "_");
    default:
      // Never call if the type is valid
      throw new Error(`Unexpected encoding: ${ec}`);
  }
}

export function jwkThumbprint<Ec extends Encodings>(jwk: JsonWebKey & {kty: "RSA" | "EC" | "oct"}, hashAlg: HashAlg): Uint8Array;
export function jwkThumbprint<Ec extends Encodings>(jwk: JsonWebKey, hashAlg: HashAlg): Uint8Array | undefined;

/**
 * Calculate JWK Thumbprint
 *
 * https://tools.ietf.org/html/rfc7638#section-3.1
 * @param jwk
 * @param hashAlg
 */
export function jwkThumbprint(jwk: JsonWebKey, hashAlg: HashAlg): Uint8Array | undefined {
  return jwkThumbprintByEncoding(jwk, hashAlg, 'uint8array');
}
